sound/soc/sof/pm.c

```diff
@@ -136,6 +136,9 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 		if (!sof_ops(sdev)->suspend)
 			return 0;
 
+	/* remove pipelines */
```

Comment:
so this one would be a fix of your previous work, and should be bundled with it rather than the byt pm fixes, right?

Reply:
@plbossart yes, it should have been there too but it is all the more necessary for BYT/CHT. What's happening with BYT/CHT is that when performing runtime suspend, the DSP doesnt really get powered off and if I dont destroy the pipelines, runtime resume fails because we restore the pipelines in the resume callback.

Reply:
do you mean to say that on BYT/CHT, the firmware still has all the context and when sending the restore pipeline IPCs you get error messages?

It seems like a case of not sending the IPC if they are not needed, instead of destroying the pipelines and having to recreate them again?

Reply:
@plbossart that is true. But is it acceptable that on BYT/CHT the DSP is not really going to be in D3 when we runtime suspend and will only be done during system suspend?

Reply:
@ranj063 whether you destroy the pipelines or not, the DSP will not be power gated on BYT/CHT. That's hardware, nothing we can do. In theory clock gating is still there but the leakage can't be controlled by power switches.

 I know that on some Android platforms S0ix was enabled, but there are a ton of dependencies on PMC/BIOS, at this point it's not worth it. The legacy firmware doesn't do anything either for D3.

Reply:
overall the fix should really be to deal with the topology only at boot and system resume for BYT/CHT, and for other devices in addition for each pm_runtime resume

Reply:
> I know that on some Android platforms S0ix was enabled, but there are a ton of dependencies on PMC/BIOS, at this point it's not worth it. The legacy firmware doesn't do anything either for D3.

@plbossart OK then should we just not bother enabling runtime PM for BYT/CHT at all?

Reply:
> overall the fix should really be to deal with the topology only at boot and system resume for BYT/CHT, and for other devices in addition for each pm_runtime resume

@plbossart what happens when we separate the audio client device from the core? We are going to need to take down the pipelines during suspend and restore them during resume right?

Reply:
> > I know that on some Android platforms S0ix was enabled, but there are a ton of dependencies on PMC/BIOS, at this point it's not worth it. The legacy firmware doesn't do anything either for D3.
> 
> @plbossart OK then should we just not bother enabling runtime PM for BYT/CHT at all?

there's really not much value indeed. No one will see the difference.

Reply:
> > overall the fix should really be to deal with the topology only at boot and system resume for BYT/CHT, and for other devices in addition for each pm_runtime resume
> 
> @plbossart what happens when we separate the audio client device from the core? We are going to need to take down the pipelines during suspend and restore them during resume right?

it's a good question - as in I don't know the answer.

we have a similar case with SoundWire where sometimes the parent keeps the context and sometimes it doesn't, and we pass this information to the children. So maybe on resume the child device need to be told that all context was lost and that it needs to recreate the whole shebang.